modules/storage/storage.go

```diff
@@ -0,0 +1,240 @@
+package storage
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	giteafs "code.gitea.io/gitea/modules/fs"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+
+	"gocloud.dev/blob"
+
+	// Google, Azure and S3 packages for bucket storage
+	_ "gocloud.dev/blob/azureblob"
+	_ "gocloud.dev/blob/fileblob"
+	_ "gocloud.dev/blob/gcsblob"
+	_ "gocloud.dev/blob/s3blob"
+)
+
+// FileStorage contains necessary info for reading or writing file to bucket
+type FileStorage struct {
```

Comment:
How about add some wrapper i.e. `func AvatarFileStorage(fileName string) *FileStorage`.

Reply:
@lunny sounds good, but I think we can simplify to:

 - `storage.Avatar`
 - `storage.RepoAvatar`
 - `storage.Attachment`
 - `storage.LFS`

Reply:
@cuonglm That's OK. If you use variables than functions, you should have a `Init` function to set `ctx` and the dir path.

Reply:
@lunny Ah no, I mean something like `storage.Avatar(ctx context.Context, filename string) *FileStorage)`, see my latest commits.